### ISSUE TITLE
tf.keras.experimental.export_saved_model in multi-gpu mode doesn't work

### ISSUE THREAD

**Issue by Oktai15** [tensorflow/tensorflow#30251]
I use TF 2.0. Currently, `tf.keras.experimental.export_saved_model` doesn't work in multi-gpu mode, i.e when model is declarated in `strategy.scope()`. I got the following error:

```
File .../tensorflow/python/ops/variables.py", line 514, in synchronization
    raise NotImplementedError
``` 

I tried the next code:

```
tf.keras.experimental.export_saved_model(
    model, file_path,
    serving_only=True,
    input_signature=[tf.TensorSpec(shape=[None, None, None, 3], dtype=tf.float32)]
)
```

**Comment by saberkun** [comment]
There is a new standard api to use keras.model.save() to export TF savedmodel. That worked for my case. Could you check if that api is in this release? If so, try it out. The export_saved_model api should be replaced by that.

**Comment by Oktai15** [comment]
@saberkun you mean `tf.keras.models.save_model`? Please, specify method that you mentioned more precisely.

Also, I'd like to say that I need `serving_only` and `input_signature`.

**Comment by Oktai15** [comment]
Just `model.save()` is not enough in my case, because I need to specify `serving_only` and `input_signature`

**Comment by saberkun** [comment]
@seemuch and @k-w-w 
This is also the confusion I had. Can keras team clarify with some HowTo for the serving only case?
If my model has dropout, what could be expected?

@Oktai15 
One workaround should work is: (1) save model with model.save_weights() (2) create a model without ds scope and load_weights(). Then, export_saved_model. (Like the version before this commit: https://github.com/tensorflow/models/commit/a1c47f2845d186e2bc65d3369525e5f7436519c2#diff-e547c86e1d6f441b9e652c8905ae8d62)

**Comment by Oktai15** [comment]
@saberkun I use `tf.keras.experimental.export_saved_model`, because it's the only one function in which I can specify `input_shape` for my model that I defined due to Imperative API (subclassed tf.keras.Model). It would be very sad, if you had remove it without alternative.

**Comment by k-w-w** [comment]
I believe this issue has been fixed in `model.save` by @seemuch (@seemuch do you have the link to the commit with the fix?).

**Comment by gadagashwini** [comment]
Hi @Oktai15, Looks like this has been fixed in the latest Tensorflow version.
Please find the working sample code below.

```
import tensorflow as tf
dataset1 = tf.data.Dataset.from_tensor_slices([[10, 11], [12, 13], [14, 15], [16, 17]])
dataset2 = tf.data.Dataset.from_tensor_slices([[21, 22], [23, 24], [25, 26], [27, 28]])
sample_dataset = tf.data.experimental.sample_from_datasets(
  [dataset1, dataset2], weights=[0.5, 0.5], seed=43)

my_strategy = tf.distribute.MirroredStrategy(["GPU:0", "CPU:0"])
with my_strategy.scope():
  @tf.function
  def distribute_train_epoch(dataset):
    for x in dataset:
      my_strategy.run(print, args=(x,))
  # Create a tf.keras model.
  model = tf.keras.Sequential()
  model.add(tf.keras.layers.Dense(1, input_shape=[10]))
  model.summary()

    # Save the tf.keras model in the SavedModel format.
  path = 'simple_keras_model'
  tf.compat.v1.keras.experimental.export_saved_model(model, path)
  dist_dataset = my_strategy.experimental_distribute_dataset(sample_dataset)

for _ in range(2):
  print("------------------")
  distribute_train_epoch(dist_dataset)
```
**Output**
```
Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
=================================================================
 dense (Dense)               (None, 1)                 11        
                                                                 
=================================================================
Total params: 11
Trainable params: 11
Non-trainable params: 0
_________________________________________________________________
WARNING:tensorflow:From /usr/local/lib/python3.7/dist-packages/tensorflow/python/saved_model/signature_def_utils_impl.py:204: build_tensor_info (from tensorflow.python.saved_model.utils_impl) is deprecated and will be removed in a future version.
Instructions for updating:
This function will only be available through the v1 compatibility library as tf.compat.v1.saved_model.utils.build_tensor_info or tf.compat.v1.saved_model.build_tensor_info.
INFO:tensorflow:Signatures INCLUDED in export for Classify: None
INFO:tensorflow:Signatures INCLUDED in export for Regress: None
INFO:tensorflow:Signatures INCLUDED in export for Predict: ['serving_default']
INFO:tensorflow:Signatures INCLUDED in export for Train: None
INFO:tensorflow:Signatures INCLUDED in export for Eval: None
INFO:tensorflow:No assets to save.
INFO:tensorflow:No assets to write.
INFO:tensorflow:SavedModel written to: si_keras_model/saved_model.pb
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:20: UserWarning: `tf.keras.experimental.export_saved_model` is deprecatedand will be removed in a future version. Please use `model.save(..., save_format="tf")` or `tf.keras.models.save_model(..., save_format="tf")`.
/usr/lib/python3.7/contextlib.py:112: UserWarning: `tf.keras.backend.learning_phase_scope` is deprecated and will be removed after 2020-10-11. To update it, simply pass a True/False value to the `training` argument of the `__call__` method of your layer or model.
  return next(self.gen)
/usr/local/lib/python3.7/dist-packages/keras/engine/training_v1.py:2079: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  updates=self.state_updates,
/usr/local/lib/python3.7/dist-packages/keras/saving/saved_model_experimental.py:284: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  g.get_collection_ref(tf.compat.v1.GraphKeys.UPDATE_OPS).extend(clone.state_updates)
------------------
/usr/local/lib/python3.7/dist-packages/tensorflow/python/data/ops/dataset_ops.py:449: UserWarning: To make it possible to preserve tf.data options across serialization boundaries, their implementation has moved to be part of the TensorFlow graph. As a consequence, the options value is in general no longer known at graph construction time. Invoking this method in graph mode retains the legacy behavior of the original implementation, but note that the returned value might not reflect the actual value of the options.
  warnings.warn("To make it possible to preserve tf.data options across "
[10]
[11]
[21]
[22]
[23]
[24]
[25]
[26]
[12]
[13]
[14]
[15]
[27]
[28]
[16]
[17]
------------------
[10]
[11]
[21]
[22]
[23]
[24]
[25]
[26]
[12]
[13]
[14]
[15]
[27]
[28]
[16]
[17]
```
